automation_tools/__init__.py

```diff
@@ -3067,61 +3070,84 @@ def download_manifest(url=None, consumer=None):
+                     '?lazy_regen=false').format(
+                     base64string.decode('utf-8'), url, consumer)
+        run(certs_put)
+
+        command = ('curl -sk -H "Authorization:Basic {0}"'
+                   ' {1}/subscription/consumers/{2}/export/').format(
+                   base64string.decode('utf-8'), url, consumer)
+
+        response = run(command + ' -I')
+        if ('Content-Disposition: attachment' in response):
+            run(command + ' -o {0}'.format(manifest_file))
+            return manifest_file
+        else:
+            raise ValueError('Request has returned no attachment. Check the'
+                             ' session and distributor hash')
 
 
 def validate_manifest(manifest_file):
     """Make sure that manifest contains only subscriptions specified in config file
-    specified in environment variable:
-
+    specified in environment variable and attach if missing any:
     EXP_SUBS_FILE
         List of subscriptions which are supposed to be in the manifest (one
         subscription name per line). No other subscription is allowed.
-
     :param manifest_file: Where is the manifest to investigate.
     """
     exp_subs_file = os.environ.get('EXP_SUBS_FILE', None)
     with open(exp_subs_file, 'r') as fp:
-        exp_subs = set([r.strip() for r in fp.readlines() if not r.strip().startswith('#')])
-    print("expected subscriptions in %s are %s" % (manifest_file, ' & '.join(exp_subs)))
+        exp_subs = {}
+        exp_subs_details = [r.strip() for r in fp.readlines() if not r.strip().startswith('#')]
+        for name in exp_subs_details:
+            exp_subs[name.split(';')[0]] = [name.split(';')[1], name.split(';')[2]]
+
+    print("expected subscriptions in %s are %s" % (manifest_file, ' & '.join(exp_subs.keys())))
     rct_output = run("rct cat-manifest --no-content %s" % manifest_file).split("\n")
     sub_name_next = False
-    current_subs = set()
+    current_subs = list()
     for r in rct_output:
         r = r.strip()
         if r == 'Subscription:':
             sub_name_next = True
             continue
         if sub_name_next:
             if r.startswith('Name: '):
-                current_subs.add(r[6:])
+                current_subs.append(r[6:])
             sub_name_next = False
     print("current subscriptions in %s are %s" % (manifest_file, ' & '.join(current_subs)))
-    if exp_subs == current_subs:
-        return True
-    else:
-        return False
+
+    response = [True]
+    for sub in exp_subs.keys():
+        if sub in current_subs:
+            continue
+        else:
+            response.append('Yes')
+            print("Attaching missing subscription %s" % (sub))
+            attach = download_manifest(pool_id=exp_subs[sub][0], count=exp_subs[sub][1])
+            if attach:
+                print("Successfully attached subscription %s" % (sub))
+            else:
+                return [False]
+    return response
```

Comment:
I dislike this complex undocumented return value of the function. If we would have separate function to fix entitlements, this could stay very simple. Or at least return touple instead of list?

Reply:
Move the from list to tuple in return value

Reply:
@jhutar I used datacollection type list  because :
- I found it simple to use for scenario where len is helping to validate.
- True/False are using based on condition and helping to pass/fail assert too.

I agree with you it's not added in docstring and I could add it now if you are fine with it ?

Reply:
Yes, if you could add it, that would help. Thank you!

Reply:
Done.